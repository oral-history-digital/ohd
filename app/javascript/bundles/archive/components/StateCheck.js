import React from 'react';
import PropTypes from 'prop-types';
import { useSelector } from 'react-redux';

import Spinner from './Spinner';

export default function StateCheck({
    testSelector,
    fallback = <Spinner />,
    children,
}) {
    const testResult = useSelector(testSelector);

    if (!testResult) {
        return fallback;
    }

    return children;
}

StateCheck.propTypes = {
    testSelector: PropTypes.func.isRequired,
    fallback: PropTypes.element,
    children: PropTypes.oneOfType([
        PropTypes.arrayOf(PropTypes.node),
        PropTypes.node,
    ]),
};
