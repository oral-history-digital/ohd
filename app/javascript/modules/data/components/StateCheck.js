import PropTypes from 'prop-types';
import { useSelector } from 'react-redux';

import { Spinner } from 'modules/spinners';

export default function StateCheck({
    testSelector,
    fallback = <Spinner />,
    children,
}) {
    const testResult = useSelector(testSelector);

    if (!testResult) {
        return fallback;
    }

    return children;
}

StateCheck.propTypes = {
    testSelector: PropTypes.func.isRequired,
    fallback: PropTypes.element,
    children: PropTypes.oneOfType([
        PropTypes.arrayOf(PropTypes.node),
        PropTypes.node,
    ]),
};
