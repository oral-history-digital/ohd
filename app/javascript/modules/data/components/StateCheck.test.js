import renderer from 'react-test-renderer';

import StateCheck from './StateCheck';

const mockedState = {
    isTrue: true,
    isFalse: false,
};

jest.mock('react-redux', () => ({
    ...jest.requireActual('react-redux'),
    useSelector: jest.fn()
      .mockImplementation(callback => callback(mockedState))
  }));

it('renders the children if selector returns true', () => {
    const selector = state => state.isTrue;

    const tree = renderer
        .create(
            <StateCheck testSelector={selector}>
                <p>Hello!</p>
            </StateCheck>
        )
        .toJSON();
    expect(tree).toMatchSnapshot();
});

it('renders Spinner component if selector returns false', () => {
    const selector = state => state.isFalse;

    const tree = renderer
        .create(
            <StateCheck testSelector={selector}>
                <p>won't be shown</p>
            </StateCheck>
        )
        .toJSON();
    expect(tree).toMatchSnapshot();
});

it('renders custom fallback', () => {
    const selector = state => state.isFalse;

    const tree = renderer
        .create(
            <StateCheck
                testSelector={selector}
                fallback={<p>Loading...</p>}
            >
                <p>won't be shown</p>
            </StateCheck>
        )
        .toJSON();
    expect(tree).toMatchSnapshot();
});
