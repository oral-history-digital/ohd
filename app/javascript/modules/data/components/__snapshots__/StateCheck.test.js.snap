// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Spinner component if selector returns false 1`] = `
<img
  alt=""
  className="Spinner"
  src="mock.jpg"
/>
`;

exports[`renders custom fallback 1`] = `
<p>
  Loading...
</p>
`;

exports[`renders the children if selector returns true 1`] = `
<p>
  Hello!
</p>
`;
